In [74]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
import re

In [36]:
movies = pd.read_csv('.\Dataset\data\data_joined.csv')

<>:1: SyntaxWarning: invalid escape sequence '\D'
<>:1: SyntaxWarning: invalid escape sequence '\D'
C:\Users\Aniket\AppData\Local\Temp\ipykernel_1992\1183933192.py:1: SyntaxWarning: invalid escape sequence '\D'
  movies = pd.read_csv('.\Dataset\data\data_joined.csv')


In [37]:
movies.head()

movie_name  Movie_ID        Budget  \
0  Star Wars: Episode VII - The Force Awakens   2488496  $245,000,000   
1                                   Frozen II   4520988  $150,000,000   
2                       The Dark Knight Rises   1345836  $250,000,000   
3                        Beauty and the Beast   2771200  $160,000,000   
4                                Finding Dory   2277860  $200,000,000   

                                                Cast  \
0  Daisy Ridley,John Boyega,Oscar Isaac,Domhnall ...   
1  Kristen Bell,Idina Menzel,Josh Gad,Jonathan Gr...   
2  Christian Bale,Tom Hardy,Anne Hathaway,Gary Ol...   
3  Emma Watson,Dan Stevens,Luke Evans,Josh Gad,Ke...   
4  Ellen DeGeneres,Albert Brooks,Ed O'Neill,Kaitl...   

                                              Writer  \
0          Lawrence Kasdan,Michael Arndt,J.J. Abrams   
1    Jennifer Lee,Hans Christian Andersen,Chris Buck   
2    Jonathan Nolan,Christopher Nolan,David S. Goyer   
3  Evan Spiliotopoulos,Bill Condon,Stephen Chbosk...   
4      Angus MacLane,Victoria Strouse,Andrew Stanton   

                                             Studios  \
0          Lucasfilm,Bad Robot,Truenorth Productions   
1  Walt Disney Animation Studios,Walt Disney Pict...   
2  Warner Bros.,Legendary Entertainment,DC Entert...   
3              Mandeville Films,Walt Disney Pictures   
4       Pixar Animation Studios,Walt Disney Pictures   

                                               Genre  \
0                            Action,Adventure,Sci-Fi   
1  Animation,Adventure,Comedy,Family,Fantasy,Musical   
2                                 Action,Crime,Drama   
3           Adventure,Family,Fantasy,Musical,Romance   
4                  Animation,Adventure,Comedy,Family   

                                            Keywords           Languages  \
0  reboot,sanitation employee,remake,crash landin...             English   
1  autumn,anthropomorphic snowman,princess,disney...             English   
2  dc comics,batman character,bruce wayne charact...      English,Arabic   
3  beast,fairy tale,heroine,beast's heart,remake ...             English   
4  fish,ocean,whale,octopus driving a truck,talki...  English,Indonesian   

                      Countries  \
0                 United States   
1                 United States   
2  United Kingdom,United States   
3                 United States   
4                 United States   

                                    Filming_Location Release_Data  \
0             Skellig Michael, County Kerry, Ireland   2015-12-18   
1                                      Norway, Norge   2019-11-22   
2         Mehrangarh Fort, Jodhpur, Rajasthan, India   2012-07-27   
3  Shepperton Studios, Studios Road, Shepperton, ...   2017-03-17   
4  Pixar Animation Studios - 1200 Park Avenue, Em...   2016-06-17   

              Runtime Gross_worldwide  Rating Rating_Count ListOfCertificate  
0  2 hours 18 minutes  $2,069,521,700     7.8         893K             PG-13  
1   1 hour 43 minutes  $1,450,026,933     6.8         156K                PG  
2  2 hours 44 minutes  $1,081,142,612     8.4         1.6M             PG-13  
3   2 hours 9 minutes  $1,273,576,220     7.1         293K                PG  
4   1 hour 37 minutes  $1,028,570,942     7.3         259K                PG

In [38]:
print("Na value stats:\n",movies.isna().sum())

Na value stats:
 movie_name              0
Movie_ID                0
Budget               2930
Cast                    9
Writer                  4
Studios                49
Genre                1178
Keywords               27
Languages              36
Countries               6
Filming_Location      493
Release_Data            0
Runtime                 7
Gross_worldwide      1035
Rating                  7
Rating_Count            7
ListOfCertificate    1199
dtype: int64


In [39]:
print("Null value stats: \n", movies.isnull().sum())

Null value stats: 
 movie_name              0
Movie_ID                0
Budget               2930
Cast                    9
Writer                  4
Studios                49
Genre                1178
Keywords               27
Languages              36
Countries               6
Filming_Location      493
Release_Data            0
Runtime                 7
Gross_worldwide      1035
Rating                  7
Rating_Count            7
ListOfCertificate    1199
dtype: int64


In [40]:
print("Duplicated value stats: ", movies.duplicated().sum())

Duplicated value stats:  0


In [41]:
movies.duplicated().sum()

np.int64(0)

In [42]:
movies.shape

(9787, 17)

Columns having significant missing values are.

Total movies in the dataset: 9787

- Budget (2930)  :: Reason 1: Since, this field is specific to each movie, we cannot assign any mean or median value to it. We just have to drop 
                    these missing fields
- Genre (1178)   :: Reason 1
- Filming_location (493)  :: Reason 2: This feature is not very important (As of now, Conclusion made on basis of general knowledge)
- Gross_worldwide (1035)   :: Reason 1 
- List_of_certificate (1199) :: Reason 2

In [43]:
# movies.drop(columns=['Filming_Location', 'ListOfCertificate'], inplace=True)

In [44]:
movies.shape

(9787, 17)

Now, checking how many missing values are still here

In [45]:
movies.isna().sum()

movie_name              0
Movie_ID                0
Budget               2930
Cast                    9
Writer                  4
Studios                49
Genre                1178
Keywords               27
Languages              36
Countries               6
Filming_Location      493
Release_Data            0
Runtime                 7
Gross_worldwide      1035
Rating                  7
Rating_Count            7
ListOfCertificate    1199
dtype: int64

In [46]:
movies.dropna(subset=['Budget'], inplace=True)

In [47]:
movies.shape

(6857, 17)

In [48]:
movies.isna().sum()

movie_name             0
Movie_ID               0
Budget                 0
Cast                   1
Writer                 1
Studios               10
Genre                639
Keywords               9
Languages             18
Countries              0
Filming_Location     208
Release_Data           0
Runtime                2
Gross_worldwide      574
Rating                 5
Rating_Count           5
ListOfCertificate    532
dtype: int64

In [49]:
movies.dropna( inplace=True)

In [50]:
movies.shape

(5404, 17)

In [51]:
# Now, Since, the data is now cleaned, we can look for the relevant features for the model.

In [52]:
movies.describe()

Movie_ID       Rating
count  5.404000e+03  5404.000000
mean   9.091608e+05     6.398057
std    1.434607e+06     0.985556
min    1.857800e+04     1.400000
25%    1.066548e+05     5.800000
50%    2.787770e+05     6.500000
75%    1.175814e+06     7.100000
max    9.844522e+06     9.200000

In [54]:
# Also, the movie title and ID won't be contributing the model prediction so we can drop them.
movies.drop(columns=['movie_name', 'Movie_ID'], inplace=True)

In [55]:
movies.shape

(5404, 15)

In [56]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5404 entries, 0 to 9786
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Budget             5404 non-null   object 
 1   Cast               5404 non-null   object 
 2   Writer             5404 non-null   object 
 3   Studios            5404 non-null   object 
 4   Genre              5404 non-null   object 
 5   Keywords           5404 non-null   object 
 6   Languages          5404 non-null   object 
 7   Countries          5404 non-null   object 
 8   Filming_Location   5404 non-null   object 
 9   Release_Data       5404 non-null   object 
 10  Runtime            5404 non-null   object 
 11  Gross_worldwide    5404 non-null   object 
 12  Rating             5404 non-null   float64
 13  Rating_Count       5404 non-null   object 
 14  ListOfCertificate  5404 non-null   object 
dtypes: float64(1), object(14)
memory usage: 675.5+ KB


In [57]:
movies.head()

Budget                                               Cast  \
0  $245,000,000  Daisy Ridley,John Boyega,Oscar Isaac,Domhnall ...   
1  $150,000,000  Kristen Bell,Idina Menzel,Josh Gad,Jonathan Gr...   
2  $250,000,000  Christian Bale,Tom Hardy,Anne Hathaway,Gary Ol...   
3  $160,000,000  Emma Watson,Dan Stevens,Luke Evans,Josh Gad,Ke...   
4  $200,000,000  Ellen DeGeneres,Albert Brooks,Ed O'Neill,Kaitl...   

                                              Writer  \
0          Lawrence Kasdan,Michael Arndt,J.J. Abrams   
1    Jennifer Lee,Hans Christian Andersen,Chris Buck   
2    Jonathan Nolan,Christopher Nolan,David S. Goyer   
3  Evan Spiliotopoulos,Bill Condon,Stephen Chbosk...   
4      Angus MacLane,Victoria Strouse,Andrew Stanton   

                                             Studios  \
0          Lucasfilm,Bad Robot,Truenorth Productions   
1  Walt Disney Animation Studios,Walt Disney Pict...   
2  Warner Bros.,Legendary Entertainment,DC Entert...   
3              Mandeville Films,Walt Disney Pictures   
4       Pixar Animation Studios,Walt Disney Pictures   

                                               Genre  \
0                            Action,Adventure,Sci-Fi   
1  Animation,Adventure,Comedy,Family,Fantasy,Musical   
2                                 Action,Crime,Drama   
3           Adventure,Family,Fantasy,Musical,Romance   
4                  Animation,Adventure,Comedy,Family   

                                            Keywords           Languages  \
0  reboot,sanitation employee,remake,crash landin...             English   
1  autumn,anthropomorphic snowman,princess,disney...             English   
2  dc comics,batman character,bruce wayne charact...      English,Arabic   
3  beast,fairy tale,heroine,beast's heart,remake ...             English   
4  fish,ocean,whale,octopus driving a truck,talki...  English,Indonesian   

                      Countries  \
0                 United States   
1                 United States   
2  United Kingdom,United States   
3                 United States   
4                 United States   

                                    Filming_Location Release_Data  \
0             Skellig Michael, County Kerry, Ireland   2015-12-18   
1                                      Norway, Norge   2019-11-22   
2         Mehrangarh Fort, Jodhpur, Rajasthan, India   2012-07-27   
3  Shepperton Studios, Studios Road, Shepperton, ...   2017-03-17   
4  Pixar Animation Studios - 1200 Park Avenue, Em...   2016-06-17   

              Runtime Gross_worldwide  Rating Rating_Count ListOfCertificate  
0  2 hours 18 minutes  $2,069,521,700     7.8         893K             PG-13  
1   1 hour 43 minutes  $1,450,026,933     6.8         156K                PG  
2  2 hours 44 minutes  $1,081,142,612     8.4         1.6M             PG-13  
3   2 hours 9 minutes  $1,273,576,220     7.1         293K                PG  
4   1 hour 37 minutes  $1,028,570,942     7.3         259K                PG

In [58]:
# # Other pre processing steps, 
# 1. remove the dollar sign in Budget and conver the column to numerical field.
# 2. conver the release date to datetime field.
# 3. Convert the runtime to time field(minutes).
# 4. Convert rating count to numerical field. 
# 5. make bins of the rating field to create the target variable field of HIT, AVERAGE, FLOP.
# 6. remove the unnamed field from the begining of the dataset.

# and name the final dataset as movies_cleaned.csv

In [61]:
movies['Budget'] = movies['Budget'].replace({'\$': '', '€': '', '£': '', ',': ''}, regex=True).astype(float)
movies.head()

<>:1: SyntaxWarning: invalid escape sequence '\$'
<>:1: SyntaxWarning: invalid escape sequence '\$'
C:\Users\Aniket\AppData\Local\Temp\ipykernel_1992\3286310489.py:1: SyntaxWarning: invalid escape sequence '\$'
  movies['Budget'] = movies['Budget'].replace({'\$': '', '€': '', '£': '', ',': ''}, regex=True).astype(float)


Budget                                               Cast  \
0  245000000.0  Daisy Ridley,John Boyega,Oscar Isaac,Domhnall ...   
1  150000000.0  Kristen Bell,Idina Menzel,Josh Gad,Jonathan Gr...   
2  250000000.0  Christian Bale,Tom Hardy,Anne Hathaway,Gary Ol...   
3  160000000.0  Emma Watson,Dan Stevens,Luke Evans,Josh Gad,Ke...   
4  200000000.0  Ellen DeGeneres,Albert Brooks,Ed O'Neill,Kaitl...   

                                              Writer  \
0          Lawrence Kasdan,Michael Arndt,J.J. Abrams   
1    Jennifer Lee,Hans Christian Andersen,Chris Buck   
2    Jonathan Nolan,Christopher Nolan,David S. Goyer   
3  Evan Spiliotopoulos,Bill Condon,Stephen Chbosk...   
4      Angus MacLane,Victoria Strouse,Andrew Stanton   

                                             Studios  \
0          Lucasfilm,Bad Robot,Truenorth Productions   
1  Walt Disney Animation Studios,Walt Disney Pict...   
2  Warner Bros.,Legendary Entertainment,DC Entert...   
3              Mandeville Films,Walt Disney Pictures   
4       Pixar Animation Studios,Walt Disney Pictures   

                                               Genre  \
0                            Action,Adventure,Sci-Fi   
1  Animation,Adventure,Comedy,Family,Fantasy,Musical   
2                                 Action,Crime,Drama   
3           Adventure,Family,Fantasy,Musical,Romance   
4                  Animation,Adventure,Comedy,Family   

                                            Keywords           Languages  \
0  reboot,sanitation employee,remake,crash landin...             English   
1  autumn,anthropomorphic snowman,princess,disney...             English   
2  dc comics,batman character,bruce wayne charact...      English,Arabic   
3  beast,fairy tale,heroine,beast's heart,remake ...             English   
4  fish,ocean,whale,octopus driving a truck,talki...  English,Indonesian   

                      Countries  \
0                 United States   
1                 United States   
2  United Kingdom,United States   
3                 United States   
4                 United States   

                                    Filming_Location Release_Data  \
0             Skellig Michael, County Kerry, Ireland   2015-12-18   
1                                      Norway, Norge   2019-11-22   
2         Mehrangarh Fort, Jodhpur, Rajasthan, India   2012-07-27   
3  Shepperton Studios, Studios Road, Shepperton, ...   2017-03-17   
4  Pixar Animation Studios - 1200 Park Avenue, Em...   2016-06-17   

              Runtime Gross_worldwide  Rating Rating_Count ListOfCertificate  
0  2 hours 18 minutes  $2,069,521,700     7.8         893K             PG-13  
1   1 hour 43 minutes  $1,450,026,933     6.8         156K                PG  
2  2 hours 44 minutes  $1,081,142,612     8.4         1.6M             PG-13  
3   2 hours 9 minutes  $1,273,576,220     7.1         293K                PG  
4   1 hour 37 minutes  $1,028,570,942     7.3         259K                PG

In [62]:
movies['Release_Data'] = pd.to_datetime(movies['Release_Data'])

In [63]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5404 entries, 0 to 9786
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Budget             5404 non-null   float64       
 1   Cast               5404 non-null   object        
 2   Writer             5404 non-null   object        
 3   Studios            5404 non-null   object        
 4   Genre              5404 non-null   object        
 5   Keywords           5404 non-null   object        
 6   Languages          5404 non-null   object        
 7   Countries          5404 non-null   object        
 8   Filming_Location   5404 non-null   object        
 9   Release_Data       5404 non-null   datetime64[ns]
 10  Runtime            5404 non-null   object        
 11  Gross_worldwide    5404 non-null   object        
 12  Rating             5404 non-null   float64       
 13  Rating_Count       5404 non-null   object        
 14  ListOfCertifi

In [75]:
def convert_runtime(runtime):
    hours = 0
    minutes = 0
    
    # Use regular expressions to capture hours and minutes
    hour_match = re.search(r'(\d+)\s*hour', runtime)
    minute_match = re.search(r'(\d+)\s*minute', runtime)
    
    # If an hour part is found, convert it to minutes
    if hour_match:
        hours = int(hour_match.group(1))
        
    # If a minute part is found, add the minutes
    if minute_match:
        minutes = int(minute_match.group(1))
    
    # Convert hours to minutes and return total
    return hours * 60 + minutes

# Apply the conversion function to the 'Runtime' column
movies['Runtime'] = movies['Runtime'].apply(convert_runtime)


In [76]:
bins = [0, 5, 7, 10]  # FLOP: 0-5, AVERAGE: 5-7, HIT: 7-10
labels = ['FLOP', 'AVERAGE', 'HIT']
movies['Target'] = pd.cut(movies['Rating'], bins=bins, labels=labels)

In [77]:
movies = movies.loc[:, ~movies.columns.str.contains('^Unnamed')]

In [ ]:
def convert_rating_count(value):
    if 'K' in value:
        return float(value.replace('K', '')) * 1_000
    elif 'M' in value:
        return float(value.replace('M', '')) * 1_000_000
    else:
        return float(value)

movies['Rating_Count'] = movies['Rating_Count'].apply(convert_rating_count)

In [84]:
movies.head()

Budget                                               Cast  \
0  245000000.0  Daisy Ridley,John Boyega,Oscar Isaac,Domhnall ...   
1  150000000.0  Kristen Bell,Idina Menzel,Josh Gad,Jonathan Gr...   
2  250000000.0  Christian Bale,Tom Hardy,Anne Hathaway,Gary Ol...   
3  160000000.0  Emma Watson,Dan Stevens,Luke Evans,Josh Gad,Ke...   
4  200000000.0  Ellen DeGeneres,Albert Brooks,Ed O'Neill,Kaitl...   

                                              Writer  \
0          Lawrence Kasdan,Michael Arndt,J.J. Abrams   
1    Jennifer Lee,Hans Christian Andersen,Chris Buck   
2    Jonathan Nolan,Christopher Nolan,David S. Goyer   
3  Evan Spiliotopoulos,Bill Condon,Stephen Chbosk...   
4      Angus MacLane,Victoria Strouse,Andrew Stanton   

                                             Studios  \
0          Lucasfilm,Bad Robot,Truenorth Productions   
1  Walt Disney Animation Studios,Walt Disney Pict...   
2  Warner Bros.,Legendary Entertainment,DC Entert...   
3              Mandeville Films,Walt Disney Pictures   
4       Pixar Animation Studios,Walt Disney Pictures   

                                               Genre  \
0                            Action,Adventure,Sci-Fi   
1  Animation,Adventure,Comedy,Family,Fantasy,Musical   
2                                 Action,Crime,Drama   
3           Adventure,Family,Fantasy,Musical,Romance   
4                  Animation,Adventure,Comedy,Family   

                                            Keywords           Languages  \
0  reboot,sanitation employee,remake,crash landin...             English   
1  autumn,anthropomorphic snowman,princess,disney...             English   
2  dc comics,batman character,bruce wayne charact...      English,Arabic   
3  beast,fairy tale,heroine,beast's heart,remake ...             English   
4  fish,ocean,whale,octopus driving a truck,talki...  English,Indonesian   

                      Countries  \
0                 United States   
1                 United States   
2  United Kingdom,United States   
3                 United States   
4                 United States   

                                    Filming_Location Release_Data  Runtime  \
0             Skellig Michael, County Kerry, Ireland   2015-12-18      138   
1                                      Norway, Norge   2019-11-22      103   
2         Mehrangarh Fort, Jodhpur, Rajasthan, India   2012-07-27      164   
3  Shepperton Studios, Studios Road, Shepperton, ...   2017-03-17      129   
4  Pixar Animation Studios - 1200 Park Avenue, Em...   2016-06-17       97   

  Gross_worldwide  Rating  Rating_Count ListOfCertificate   Target  
0  $2,069,521,700     7.8      893000.0             PG-13      HIT  
1  $1,450,026,933     6.8      156000.0                PG  AVERAGE  
2  $1,081,142,612     8.4     1600000.0             PG-13      HIT  
3  $1,273,576,220     7.1      293000.0                PG      HIT  
4  $1,028,570,942     7.3      259000.0                PG      HIT

In [85]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5404 entries, 0 to 9786
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Budget             5404 non-null   float64       
 1   Cast               5404 non-null   object        
 2   Writer             5404 non-null   object        
 3   Studios            5404 non-null   object        
 4   Genre              5404 non-null   object        
 5   Keywords           5404 non-null   object        
 6   Languages          5404 non-null   object        
 7   Countries          5404 non-null   object        
 8   Filming_Location   5404 non-null   object        
 9   Release_Data       5404 non-null   datetime64[ns]
 10  Runtime            5404 non-null   int64         
 11  Gross_worldwide    5404 non-null   object        
 12  Rating             5404 non-null   float64       
 13  Rating_Count       5404 non-null   float64       
 14  ListOfCertifi

In [86]:
movies.describe()

Budget                   Release_Data      Runtime       Rating  \
count  5.404000e+03                           5404  5404.000000  5404.000000   
mean   3.511625e+07  2001-06-11 00:46:53.915618048   109.145078     6.398057   
min    1.500000e+04            1929-01-05 00:00:00    61.000000     1.400000   
25%    1.000000e+07            1993-04-30 00:00:00    96.000000     5.800000   
50%    2.000000e+07            2002-11-27 00:00:00   106.000000     6.500000   
75%    4.200000e+07            2011-02-11 00:00:00   118.000000     7.100000   
max    3.560000e+08            2021-09-30 00:00:00   271.000000     9.200000   
std    4.170425e+07                            NaN    18.485515     0.985556   

       Rating_Count  
count  5.404000e+03  
mean   1.152212e+05  
min    1.210000e+02  
25%    1.600000e+04  
50%    5.000000e+04  
75%    1.300000e+05  
max    2.500000e+06  
std    1.883993e+05

In [87]:
movies.to_csv('.\Dataset\data\movies_cleaned.csv', index=False)

<>:1: SyntaxWarning: invalid escape sequence '\D'
<>:1: SyntaxWarning: invalid escape sequence '\D'
C:\Users\Aniket\AppData\Local\Temp\ipykernel_1992\3981318727.py:1: SyntaxWarning: invalid escape sequence '\D'
  movies.to_csv('.\Dataset\data\movies_cleaned.csv', index=False)


In [80]:
# Now, the movie is properly cleaned, 
# Next, we can move forward with the NLP task (converting the textual data into numerical representatoins)

In [88]:
## PRE PROCESSING DONE

In [89]:
## NOW, MOVE FORWARD WITH THE NLP TASK